### ISSUE TITLE
Type checking of output arguments

### ISSUE THREAD

**Issue by UWN** [ichiban/prolog#235]
Many built-in predicates require the type checking of the output argument. Conceptually, this is best represented by a checking goal like `can_be(list, Xs)`. The idea is that `can_be(Type, Term)` ensures there is an instance of `Term` such that `Term`&theta; is of type or domain `Type`. And if it is not, it produces mostly a type or domain error (7.12.2 b, c).

```
?- sort([],[t|1]).
?- a =.. [t|1].
?- term_variables(_,[t|1]).
?- findall(t,true,[t|1]).
?- bagof(t,true,[t|1]).
?- setof(t,true,[t|1]).
?- atom_chars(a, [t|1]).
   false, unexpected. % for all above
   type_error(list,[t|1]). % expected for all above.
```

And contrast this to `must_be/2` which ensures that for all instances it holds that `Term` is of type or domain `Type`.

**Comment by UWN** [comment]
Ad *mostly produces a type or domain error*:

```
?- char_code(Ch,-2).
2022/07/23 09:11:09 error(representation_error(character_code),char_code/2)
```

This is a case where instead of a type or domain error a representation error is produces. The motivation behind is that character sets often can be extended, so maybe that integer is in fact member of a larger character set, which is not supported by a processor.

From a semantic viewpoint representation errors do not produce as precise information as type or domain errors. While type and domain errors are tantamount to failure and thus (for simple built-ins that do not contain further goals) mean that there is simply no solution at all, representation errors just mean that some limit has been reached and we cannot conclude anything about the meaning of that error.

This is again another good reason to prefer` chars` over `codes`.

**Comment by ichiban** [comment]
Fixed in `v0.10.3`.

```console
$ $(go env GOPATH)/bin/1pl
Top level for ichiban/prolog v0.10.3
This is for testing purposes only!
See https://github.com/ichiban/prolog for more details.
Type Ctrl-C or 'halt.' to exit.
?- sort([],[t|1]).
2022/08/01 10:23:37 error(type_error(list,[t|1]),sort/2)
?- a =.. [t|1].
2022/08/01 10:23:44 error(type_error(list,[t|1]),=.. /2)
?- term_variables(_,[t|1]).
2022/08/01 10:23:51 error(type_error(list,[t|1]),term_variables/2)
?- findall(t,true,[t|1]).
2022/08/01 10:23:58 error(type_error(list,[t|1]),findall/3)
?- bagof(t,true,[t|1]).
2022/08/01 10:24:06 error(type_error(list,[t|1]),bagof/3)
?- setof(t,true,[t|1]).
2022/08/01 10:24:13 error(type_error(list,[t|1]),setof/3)
?- atom_chars(a, [t|1]).
2022/08/01 10:24:21 error(type_error(list,[t|1]),atom_chars/2)
?- atom_codes(a, [0't|1]).
2022/08/01 10:24:32 error(type_error(list,[116|1]),atom_codes/2)
?- 
```

**Comment by UWN** [comment]
Excellent!
